src/hotspot/share/runtime/thread.cpp

```diff
@@ -1644,7 +1646,7 @@ void JavaThread::check_and_handle_async_exceptions() {
     case _thread_in_Java: {
       ThreadInVMfromJava tiv(this);
       JavaThread* THREAD = this;
-      Exceptions::throw_unsafe_access_internal_error(THREAD, __FILE__, __LINE__, "a fault occurred in a recent unsafe memory access operation in compiled Java code");
```

Comment:
This wording is somewhat deliberate as the unsafe access if not detected synchronously with the actual memory operation - hence it is a "recent unsafe memory access operation".

Reply:
I see, maybe I should instead change the error message for the _thread_in_vm case to be "a fault occurred in a recent unsafe memory access operation"?

Reply:
IIRC it was only the _thread_in_Java case that was special because that was where JIT'd code could have a series of unsafe memory accesses, and the exception won't be thrown until we hit a specific async exception check point.

Reply:
So today we always reach the _async_unsafe_access_error case of check_and_handle_async_exceptions() with a state of _thread_in_Java, except when called from jni_check_async_exceptions() where we are in vm, so I don't know why the message refers exclusively to compiled code. I run test InternalErrorTest.java in baseline with just the interpreter("main/othervm -Xint -Xbootclasspath/a:.  -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI InternalErrorTest") and still passes. Here is the code that sets the pending unsafe access error when the thread receives SIGBUS while in Java: https://github.com/openjdk/jdk/blob/master/src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp#L249

Reply:
If the unsafe access error happens on Unsafe_CopySwapMemory0 like in the test then we will be either in vm or native (depending on src & dst) which will also output the same error message. We can differentiate between this cases if we want by saving the state in SharedRuntime::handle_unsafe_access().

Reply:
Sorry for the delay in getting back to this. There are a lot of subtleties around the throwing of an async exception related to "unsafe access errors" and I had to dig around to refresh my memory on a few things. End result: changing the error messages to say the same thing is fine.

Reply:
Thanks for the review David!